<a href="https://codeimmersives.com"><img src = "https://www.codeimmersives.com/wp-content/uploads/2019/09/CodeImmersives_Logo_RGB_NYC_BW.png" width = 400> </a>


<h1 align=center><font size = 5>Python Assessment</font></h1>

<h2 #0>Exercise 1</h2><a id="0"></a>
<code>
purchases = {'2021-01-01': 
                 {'item-001': 
                      {'name': 'Alka Seltzer'},
                 'item-002': 
                      {'name': 'Rolaids'},
                 'item-003': 
                      {'name': 'Orange juice'}  }                
            }
</code>
<br>
<h4>List all of the names of items purchased</h4>

In [10]:
purchases = {'2021-01-01': 
                 {'item-001': 
                      {'name': 'Alka Seltzer'},
                 'item-002': 
                      {'name': 'Rolaids'},
                 'item-003': 
                      {'name': 'Orange juice'}  }                
            }
print(type(purchases))
if isinstance(purchases,dict):
    print(f'I found a dictionary {purchases}')
print('='*35)
# Step 1 - iterate
for k,v in purchases.items():
    #print(v)
    for k2,v2 in v.items():
        # print(v2)
        for k3,v3 in v2.items():
            print(f'You bought {v3}')

<class 'dict'>
I found a dictionary {'2021-01-01': {'item-001': {'name': 'Alka Seltzer'}, 'item-002': {'name': 'Rolaids'}, 'item-003': {'name': 'Orange juice'}}}
You bought Alka Seltzer
You bought Rolaids
You bought Orange juice


<h2>Example of 2 keys in the dictionary with the same key name</h2>

In [1]:
toys = {'Slinky': 9.99,
       'YoYo': 14.99,
       'Slinky': 5.49}
print(toys)

{'Slinky': 5.49, 'YoYo': 14.99}


<h2>Exercise 2</h2><a id="2"></a>
<code>
purchases = {'2021-01-01': 
                 {'item-001': 
                      {'name': 'Alka Seltzer'},
                 'item-002': 
                      {'name': 'Rolaids'},
                 'item-003': 
                      {'name': 'Orange juice'}  }                
            }
</code>
<br>
1 - Add a 4th item with a key of 'item-004' and name of 'Bloody Mary'<br>
2 - Add a 5th item with a key of 'item-005' and name of 'Pineapple Juice'<br>
3 - print all item names purchased<br>
4 - print all juice items

In [67]:
purchases = {'2021-01-01': 
                 {'item-001': 
                      {'name': 'Alka Seltzer'},
                 'item-002': 
                      {'name': 'Rolaids'},
                 'item-003': 
                      {'name': 'Orange juice'}  }                
            }
# 1 - Add a 4th item with a key of 'item-004' and name of 'Bloody Mary'
purchases.keys()
purchases['2021-01-01']['item-004'] = {'name':'Bloody Mary'}
print(purchases.values())
for k,v in purchases.items():
    #print(v)
    for k2,v2 in v.items():
        # print(v2)
        for k3,v3 in v2.items():
            print(f'You bought {v3}')
print('-'*25)
# 1 - Add a 5th item with a key of 'item-005' and name of 'Pineapple Juice'
purchases.keys()
purchases['2021-01-01']['item-005'] = {'name': 'Pineapple Juice'}
for k,v in purchases.items():
    #print(v)
    for k2,v2 in v.items():
        # print(v2)
        for k3,v3 in v2.items():
            print(f'You bought {v3}')
# Print all juice items
print('Juice items-'*5)
for k,v in purchases.items():
    #print(v)
    for k2,v2 in v.items():
        # print(v2)
        for k3,v3 in v2.items():
            if 'juice' in v3.lower():
                print(f'You bought {v3}')

dict_values([{'item-001': {'name': 'Alka Seltzer'}, 'item-002': {'name': 'Rolaids'}, 'item-003': {'name': 'Orange juice'}, 'item-004': {'name': 'Bloody Mary'}}])
You bought Alka Seltzer
You bought Rolaids
You bought Orange juice
You bought Bloody Mary
-------------------------
You bought Alka Seltzer
You bought Rolaids
You bought Orange juice
You bought Bloody Mary
You bought Pineapple Juice
Juice items-Juice items-Juice items-Juice items-Juice items-
You bought Orange juice
You bought Pineapple Juice


<h2>To find all methods available to a object use dir()</h2>

In [22]:
purchases = {'2021-01-01': 
                 {'item-001': 
                      {'name': 'Alka Seltzer'},
                 'item-002': 
                      {'name': 'Rolaids'},
                 'item-003': 
                      {'name': 'Orange juice'}  }                
            }
# update
dir(purchases)

['__class__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'clear',
 'copy',
 'fromkeys',
 'get',
 'items',
 'keys',
 'pop',
 'popitem',
 'setdefault',
 'update',
 'values']

<h2>dunder methods are built in reserved methods for a class</h2>

In [29]:
# iterate over the object
purchases = {'2021-01-01': 
                 {'item-001': 
                      {'name': 'Alka Seltzer'},
                 'item-002': 
                      {'name': 'Rolaids'},
                 'item-003': 
                      {'name': 'Orange juice'}  }                
            }
methods = dir(purchases)
print(type(methods))
# iterate over a list method 1  - for loop
for i in methods:
    print(i)
print('+'*35)
# iterate over a list method 2
for i in range(len(methods)):
    print(i,methods[i])
# method 3 - include an index
print('$'*35)
for i,method_name in enumerate(methods):
    print(i,method_name)
# method 4 - list comprehension
print('X'*35)
ms = [i for i in methods ]
print(ms)


<class 'list'>
__class__
__contains__
__delattr__
__delitem__
__dir__
__doc__
__eq__
__format__
__ge__
__getattribute__
__getitem__
__gt__
__hash__
__init__
__init_subclass__
__iter__
__le__
__len__
__lt__
__ne__
__new__
__reduce__
__reduce_ex__
__repr__
__setattr__
__setitem__
__sizeof__
__str__
__subclasshook__
clear
copy
fromkeys
get
items
keys
pop
popitem
setdefault
update
values
+++++++++++++++++++++++++++++++++++
0 __class__
1 __contains__
2 __delattr__
3 __delitem__
4 __dir__
5 __doc__
6 __eq__
7 __format__
8 __ge__
9 __getattribute__
10 __getitem__
11 __gt__
12 __hash__
13 __init__
14 __init_subclass__
15 __iter__
16 __le__
17 __len__
18 __lt__
19 __ne__
20 __new__
21 __reduce__
22 __reduce_ex__
23 __repr__
24 __setattr__
25 __setitem__
26 __sizeof__
27 __str__
28 __subclasshook__
29 clear
30 copy
31 fromkeys
32 get
33 items
34 keys
35 pop
36 popitem
37 setdefault
38 update
39 values
$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
0 __class__
1 __contains__
2 __delattr__
3 __delitem__
4 __

In [34]:
# Adding a condition in our iteration
for i in methods:
    if not i[:2] == '__':
        print(i)
print('X'*35)
ms = [i for i in methods if not i[:2] == '__']
print(ms)
print('^'*35)
ms = [i for i in methods if not i.startswith('__')]
print(ms)

clear
copy
fromkeys
get
items
keys
pop
popitem
setdefault
update
values
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
['clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
['clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']


<h4>Extract all methods with the name 'item' as part of the name</h4?

In [37]:
ms = [i for i in methods if not i.startswith('__') and 'item' in i]
print(ms)

['items', 'popitem']


In [41]:
x = 100
y = 100
print(id(x),id(y))
for i in range(-10,301):
    print(i)

140723701841248 140723701841248
-10
-9
-8
-7
-6
-5
-4
-3
-2
-1
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261

In [57]:
# Which ids are the same?
# print(id(x),id(y))
# for i in range(-10,301):
#     x = int(i)
#     y = int(i)
#     #     print(i, id(x),id(y),id(x)==id(y))
x = 257
y = 257
if id(x) == id(y):
    print(f'{x},{y} The ids are the same {id(x)} {id(y)}')
else:
    print(f'{x},{y} The ids are NOT the same {id(x)} {id(y)}')    

257,257 The ids are NOT the same 1768899668048 1768899668016


In [58]:
x = [1]
y = [1]
if id(x) == id(y):
    print(f'{x},{y} The ids are the same {id(x)} {id(y)}')
else:
    print(f'{x},{y} The ids are NOT the same {id(x)} {id(y)}')  

[1],[1] The ids are NOT the same 1768899744904 1768899744648


In [62]:
x = '233'
y = '233'
if id(x) == id(y):
    print(f'{x},{y} The ids are the same {id(x)} {id(y)}')
else:
    print(f'{x},{y} The ids are NOT the same {id(x)} {id(y)}')  

233,233 The ids are the same 1768899987632 1768899987632


<h2>Exercise 3</h2><a id="4"></a>
<code>
items = ['Alka Seltzer', 'Rolaids', 'Orange juice', 'Bloody Mary', 'Pineapple Juice']
</code>
<br>
1 - Randomly pick 3 items out of the list using iteration and name it basket1 - WITHOUT REPLACEMENT<br>
2a - Randomly pick 3 items out of the list using iteration and name it basket2 - WITH REPLACEMENT<br>
2b - List the items remaining in the basket

In [76]:
import random

items = ['Alka Seltzer', 'Rolaids', 'Orange juice', 'Bloody Mary', 'Pineapple Juice']
basket1 = []
items_chosen = 0
# With replacement
for item in items:
    my_choice = random.choice(items)
    items_chosen += 1
    print(my_choice)
    basket1.append(my_choice)
    if items_chosen > 2:
        break
print(f'You chose {basket1}')
# without replacement
basket1 = []
items_chosen = 0
for item in items:
    my_choice = random.choice(items)
    items_chosen += 1
    print(my_choice)
    basket1.append(my_choice)
    items.pop(items.index(my_choice))
    if items_chosen > 2:
        break
print(f'You chose {basket1}')
print(f'The following items remained in the basket {items}')

Rolaids
Alka Seltzer
Rolaids
You chose ['Rolaids', 'Alka Seltzer', 'Rolaids']
Alka Seltzer
Orange juice
Bloody Mary
You chose ['Alka Seltzer', 'Orange juice', 'Bloody Mary']
The following items remained in the basket ['Rolaids', 'Pineapple Juice']


<h2>Exercise 4</h2><a id="6"></a>
From the numbers 1 through 120 pick all numbers that are divisible by 3 and 5

In [77]:
for i in range(1,121):
    if i % 3 == 0 and i % 5 == 0:
        print(f'{i} is divisible by 3 and 5')

15 is divisible by 3 and 5
30 is divisible by 3 and 5
45 is divisible by 3 and 5
60 is divisible by 3 and 5
75 is divisible by 3 and 5
90 is divisible by 3 and 5
105 is divisible by 3 and 5
120 is divisible by 3 and 5


<h2>Exercise 5</h2><a id="8"></a>
From the numbers 1 through 300 pick all numbers that the sum of the digits are divisible by 9<br>
For example the number 261 if we add the digits 2 + 6 + 1 = 9

In [78]:
# step 1 - iterate over the domain of the problem
for i in range(1,301):
    print(i)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


In [79]:
# step 2 - convert the integer into a str
for i in range(1,301):
    print(i, str(i))

1 1
2 2
3 3
4 4
5 5
6 6
7 7
8 8
9 9
10 10
11 11
12 12
13 13
14 14
15 15
16 16
17 17
18 18
19 19
20 20
21 21
22 22
23 23
24 24
25 25
26 26
27 27
28 28
29 29
30 30
31 31
32 32
33 33
34 34
35 35
36 36
37 37
38 38
39 39
40 40
41 41
42 42
43 43
44 44
45 45
46 46
47 47
48 48
49 49
50 50
51 51
52 52
53 53
54 54
55 55
56 56
57 57
58 58
59 59
60 60
61 61
62 62
63 63
64 64
65 65
66 66
67 67
68 68
69 69
70 70
71 71
72 72
73 73
74 74
75 75
76 76
77 77
78 78
79 79
80 80
81 81
82 82
83 83
84 84
85 85
86 86
87 87
88 88
89 89
90 90
91 91
92 92
93 93
94 94
95 95
96 96
97 97
98 98
99 99
100 100
101 101
102 102
103 103
104 104
105 105
106 106
107 107
108 108
109 109
110 110
111 111
112 112
113 113
114 114
115 115
116 116
117 117
118 118
119 119
120 120
121 121
122 122
123 123
124 124
125 125
126 126
127 127
128 128
129 129
130 130
131 131
132 132
133 133
134 134
135 135
136 136
137 137
138 138
139 139
140 140
141 141
142 142
143 143
144 144
145 145
146 146
147 147
148 148
149 149
150 150
151 151
152 152


In [81]:
# step 3 - iterate over the string
for i in range(1,301):
    total_of_digits = 0
    for digit in str(i):
        print(f'adding {digit}')
        total_of_digits += int(digit)
    print(f'The digits of {i} add up to {total_of_digits}')
    print('='*35)

adding 1
The digits of 1 add up to 1
adding 2
The digits of 2 add up to 2
adding 3
The digits of 3 add up to 3
adding 4
The digits of 4 add up to 4
adding 5
The digits of 5 add up to 5
adding 6
The digits of 6 add up to 6
adding 7
The digits of 7 add up to 7
adding 8
The digits of 8 add up to 8
adding 9
The digits of 9 add up to 9
adding 1
adding 0
The digits of 10 add up to 1
adding 1
adding 1
The digits of 11 add up to 2
adding 1
adding 2
The digits of 12 add up to 3
adding 1
adding 3
The digits of 13 add up to 4
adding 1
adding 4
The digits of 14 add up to 5
adding 1
adding 5
The digits of 15 add up to 6
adding 1
adding 6
The digits of 16 add up to 7
adding 1
adding 7
The digits of 17 add up to 8
adding 1
adding 8
The digits of 18 add up to 9
adding 1
adding 9
The digits of 19 add up to 10
adding 2
adding 0
The digits of 20 add up to 2
adding 2
adding 1
The digits of 21 add up to 3
adding 2
adding 2
The digits of 22 add up to 4
adding 2
adding 3
The digits of 23 add up to 5
adding 2

In [84]:
# step 4 - filter for items that add up to 9
for i in range(1,301):
    total_of_digits = 0
    for digit in str(i):
        # print(f'adding {digit}')
        total_of_digits += int(digit)
    if total_of_digits % 9 == 0:
        print(f'The digits of {i} add up to {total_of_digits}')
        print('='*35)

The digits of 9 add up to 9
The digits of 18 add up to 9
The digits of 27 add up to 9
The digits of 36 add up to 9
The digits of 45 add up to 9
The digits of 54 add up to 9
The digits of 63 add up to 9
The digits of 72 add up to 9
The digits of 81 add up to 9
The digits of 90 add up to 9
The digits of 99 add up to 18
The digits of 108 add up to 9
The digits of 117 add up to 9
The digits of 126 add up to 9
The digits of 135 add up to 9
The digits of 144 add up to 9
The digits of 153 add up to 9
The digits of 162 add up to 9
The digits of 171 add up to 9
The digits of 180 add up to 9
The digits of 189 add up to 18
The digits of 198 add up to 18
The digits of 207 add up to 9
The digits of 216 add up to 9
The digits of 225 add up to 9
The digits of 234 add up to 9
The digits of 243 add up to 9
The digits of 252 add up to 9
The digits of 261 add up to 9
The digits of 270 add up to 9
The digits of 279 add up to 18
The digits of 288 add up to 18
The digits of 297 add up to 18


In [88]:
total = 0
for i in range(6):
    total += i
print(total)

15


<h4>exercise extra - count backwards from 130 to 100 by 3</h4>

In [92]:
for i in range(130,100,-3):
    print(i)
    
# list comprehension
ans = [i for i in range(130,100,-3)]
print(ans)

130
127
124
121
118
115
112
109
106
103
[130, 127, 124, 121, 118, 115, 112, 109, 106, 103]


<h2>Exercise 6</h2><a id="10"></a>
<code>
numbers = [2,7,9,12,15,3,11,22,17,6,0,-8]
</code><br>
List all 2 number combinations that add up to 9<br>
List all 2 number permutations that add up to 9<br>

In [96]:
# step 1 - iterate over the domain of the object
numbers = [2,7,9,12,15,3,11,22,17,6,0,-8]
for i,num in enumerate(numbers):
    print(i,num)
print('='*35)   

# step 2 - refine my output
numbers = [2,7,9,12,15,3,11,22,17,6,0,-8]
for i,num in enumerate(numbers):
    for j,num2 in enumerate(numbers):
        print(f'index = {i} number = {num}, index = {j} number = {num2}')
# step 3 add my target
print('*'*35)
target = 9
numbers = [2,7,9,12,15,3,11,22,17,6,0,-8]
for i,num in enumerate(numbers):
    for j,num2 in enumerate(numbers):
        if not i == j and num + num2 == target:
            print(f'index = {i} number = {num}, index = {j} number = {num2}')


0 2
1 7
2 9
3 12
4 15
5 3
6 11
7 22
8 17
9 6
10 0
11 -8
index = 0 number = 2, index = 0 number = 2
index = 0 number = 2, index = 1 number = 7
index = 0 number = 2, index = 2 number = 9
index = 0 number = 2, index = 3 number = 12
index = 0 number = 2, index = 4 number = 15
index = 0 number = 2, index = 5 number = 3
index = 0 number = 2, index = 6 number = 11
index = 0 number = 2, index = 7 number = 22
index = 0 number = 2, index = 8 number = 17
index = 0 number = 2, index = 9 number = 6
index = 0 number = 2, index = 10 number = 0
index = 0 number = 2, index = 11 number = -8
index = 1 number = 7, index = 0 number = 2
index = 1 number = 7, index = 1 number = 7
index = 1 number = 7, index = 2 number = 9
index = 1 number = 7, index = 3 number = 12
index = 1 number = 7, index = 4 number = 15
index = 1 number = 7, index = 5 number = 3
index = 1 number = 7, index = 6 number = 11
index = 1 number = 7, index = 7 number = 22
index = 1 number = 7, index = 8 number = 17
index = 1 number = 7, index 

In [ ]:
# to solve for the combinations is tricky
results = []
target = 9
numbers = [2,7,9,12,15,3,11,22,17,6,0,-8]
for i,num in enumerate(numbers):
    for j,num2 in enumerate(numbers):
        if not i == j and num + num2 == target:
            if not [num,num2] in results and \
                not [num2,num] in results:
                print(f'index = {i} number = {num}, index = {j} number = {num2}')
                results.append([num,num2])
print(results)    


index = 0 number = 2, index = 1 number = 7
index = 2 number = 9, index = 10 number = 0
index = 5 number = 3, index = 9 number = 6
index = 8 number = 17, index = 11 number = -8
[[2, 7], [9, 0], [3, 6], [17, -8]]


<h2>Exercise 7</h2><a id="22"></a>
<code>
list_1 = ['chair','stools','dresser','picture frame']
list_2 = ['mirror','rug','tv stand','chair','dresser']
</code><br>
1 - Print an unduplicated list of all items in both lists (union)<br>
2 - Which items are duplicated in each list (intersection)

In [6]:
list_1 = ['chair','stools','dresser','picture frame']
list_2 = ['mirror','rug','tv stand','chair','dresser']

print (f'The lsit-1: {list_1}')
print (f'The lsit-2: {list_2}')
print()
# 1 - Print an unduplicated list of all items in both lists (union)
union_list = list(set().union(list_1,list_2))
print(f'The unduplicated list of all items in both lists: {union_list}')
# 2 - Which items are duplicated in each list (intersection)
intersection_list = list(set(list_1) & set(list_2))
print(f'The duplicated itmes in each list: {intersection_list}')

The lsit-1: ['chair', 'stools', 'dresser', 'picture frame']
The lsit-2: ['mirror', 'rug', 'tv stand', 'chair', 'dresser']

The unduplicated list of all items in both lists: ['tv stand', 'dresser', 'chair', 'picture frame', 'stools', 'rug', 'mirror']
The duplicated itmes in each list: ['chair', 'dresser']


<h2>Exercise 8</h2><a id="14"></a>
Create a simulated deck of cards as a list <br>
Cards have the following attributes:<br>
4 suits - Clubs, Spades, Diamonds and Hearts<br>
Cards numbered 2 through 10 plus Jack, Queen, King and Ace

In [161]:
import random

suits = ['Clubs', 'Spades', 'Diamonds','Hearts']
num_cards = range(2,11)
face_cards = ['Jack','Queen','King','Ace']
cards = list(num_cards) + face_cards
# print(cards,len(cards))
deck = []
for i in suits:
    for card in cards:
        deck.append(f'{card} of {i}')
print(deck)

card1 = random.choice(deck)
card2 = random.choice(deck)
print(f'You dealt 2 cards {card1} {card2}')

# shuffle my cards
new_deck = random.shuffle(deck)
print(new_deck)
help(random.shuffle)

['2 of Clubs', '3 of Clubs', '4 of Clubs', '5 of Clubs', '6 of Clubs', '7 of Clubs', '8 of Clubs', '9 of Clubs', '10 of Clubs', 'Jack of Clubs', 'Queen of Clubs', 'King of Clubs', 'Ace of Clubs', '2 of Spades', '3 of Spades', '4 of Spades', '5 of Spades', '6 of Spades', '7 of Spades', '8 of Spades', '9 of Spades', '10 of Spades', 'Jack of Spades', 'Queen of Spades', 'King of Spades', 'Ace of Spades', '2 of Diamonds', '3 of Diamonds', '4 of Diamonds', '5 of Diamonds', '6 of Diamonds', '7 of Diamonds', '8 of Diamonds', '9 of Diamonds', '10 of Diamonds', 'Jack of Diamonds', 'Queen of Diamonds', 'King of Diamonds', 'Ace of Diamonds', '2 of Hearts', '3 of Hearts', '4 of Hearts', '5 of Hearts', '6 of Hearts', '7 of Hearts', '8 of Hearts', '9 of Hearts', '10 of Hearts', 'Jack of Hearts', 'Queen of Hearts', 'King of Hearts', 'Ace of Hearts']
You dealt 2 cards Queen of Spades 5 of Hearts
None
Help on method shuffle in module random:

shuffle(x, random=None) method of random.Random instance
    

<h2>Exercise 9</h2><a id="16"></a>
a - Recreate a simulated deck of cards as a dictionary where the key is the name<br>
and the value is the value of the card.  Remember the Ace can be a 1 or 11<br>
b - Deal 3 players (player_1, player_2, player_3) 2 cards and determine who has the higher amount<br>
NOTE: Deal with the fact the Ace returns a tuple take the higher amount for now

In [174]:
import random

suits = ['Clubs', 'Spades', 'Diamonds','Hearts']
num_cards = range(2,11)
face_cards = ['Jack','Queen','King','Ace']
cards = list(num_cards) + face_cards
# method 1
card_value = [2,3,4,5,6,7,8,9,10,10,10,10,(1,11)]
# method 2 
card_value =[ x if isinstance(x,int) else \
             (1,11) if x == 'Ace' else 10 for x in cards]
print(card_value)
# print(cards)
deck = {}
for i in suits:
    for j,card in enumerate(cards):
        deck[f'{card} of {i}'] = card_value[j]
print(deck)

[2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, (1, 11)]
{'2 of Clubs': 2, '3 of Clubs': 3, '4 of Clubs': 4, '5 of Clubs': 5, '6 of Clubs': 6, '7 of Clubs': 7, '8 of Clubs': 8, '9 of Clubs': 9, '10 of Clubs': 10, 'Jack of Clubs': 10, 'Queen of Clubs': 10, 'King of Clubs': 10, 'Ace of Clubs': (1, 11), '2 of Spades': 2, '3 of Spades': 3, '4 of Spades': 4, '5 of Spades': 5, '6 of Spades': 6, '7 of Spades': 7, '8 of Spades': 8, '9 of Spades': 9, '10 of Spades': 10, 'Jack of Spades': 10, 'Queen of Spades': 10, 'King of Spades': 10, 'Ace of Spades': (1, 11), '2 of Diamonds': 2, '3 of Diamonds': 3, '4 of Diamonds': 4, '5 of Diamonds': 5, '6 of Diamonds': 6, '7 of Diamonds': 7, '8 of Diamonds': 8, '9 of Diamonds': 9, '10 of Diamonds': 10, 'Jack of Diamonds': 10, 'Queen of Diamonds': 10, 'King of Diamonds': 10, 'Ace of Diamonds': (1, 11), '2 of Hearts': 2, '3 of Hearts': 3, '4 of Hearts': 4, '5 of Hearts': 5, '6 of Hearts': 6, '7 of Hearts': 7, '8 of Hearts': 8, '9 of Hearts': 9, '10 of Hearts': 10, '

<h2>Exercise 10</h2>
Create a program the outputs the first day of the month, the last day of the month,<br>
and the number of days in the month as a list<br>
Example<br>
(1/1/2020,1/31/2020,31 days)<br>
(2/1/2020,2/29/2020,29 days)<br>
<br>
NOTE: Do not import calendar

In [175]:
# step 1 - create the months
for i in range(1,13):
    print(i)

1
2
3
4
5
6
7
8
9
10
11
12


In [176]:
import datetime

# step 2 - create the first day of every month
year = 2020
for i in range(1,13):
    first_day_of_month = datetime.date(year,i,1)
    print(first_day_of_month)
    

2020-01-01
2020-02-01
2020-03-01
2020-04-01
2020-05-01
2020-06-01
2020-07-01
2020-08-01
2020-09-01
2020-10-01
2020-11-01
2020-12-01


In [180]:
import datetime

# step 3 - get the next month minus 1 day to get the last day of that month
year = 2020
for i in range(1,13):
    first_day_of_month = datetime.date(year,i,1)
    if i == 12:
        last_day_of_month =  datetime.date(year+1,1,1) - \
            datetime.timedelta(days = 1)
    else:
        last_day_of_month =  datetime.date(year,i+1,1) - \
            datetime.timedelta(days = 1)
    print(first_day_of_month,last_day_of_month)
    

2020-01-01 2020-01-31
2020-02-01 2020-02-29
2020-03-01 2020-03-31
2020-04-01 2020-04-30
2020-05-01 2020-05-31
2020-06-01 2020-06-30
2020-07-01 2020-07-31
2020-08-01 2020-08-31
2020-09-01 2020-09-30
2020-10-01 2020-10-31
2020-11-01 2020-11-30
2020-12-01 2020-12-31


In [203]:
import datetime

# step 4 - Calculate the number of days in the month
year = 2020
for i in range(1,13):
    first_day_of_month = datetime.date(year,i,1)
    if i == 12:
        last_day_of_month =  datetime.date(year+1,1,1) - \
            datetime.timedelta(days = 1)
    else:
        last_day_of_month =  datetime.date(year,i+1,1) - \
            datetime.timedelta(days = 1)
    # method 1 - subtract the 2 dates and add 1
    # print(type(first_day_of_month),type(last_day_of_month))
    print(first_day_of_month,last_day_of_month,(last_day_of_month - \
         first_day_of_month).days+1)
    # method 2 - use slicing
    print(first_day_of_month,last_day_of_month,\
          last_day_of_month.strftime('%Y/%m/%d')[-2:])

2020-01-01 2020-01-31 31
2020-01-01 2020-01-31 31
2020-02-01 2020-02-29 29
2020-02-01 2020-02-29 29
2020-03-01 2020-03-31 31
2020-03-01 2020-03-31 31
2020-04-01 2020-04-30 30
2020-04-01 2020-04-30 30
2020-05-01 2020-05-31 31
2020-05-01 2020-05-31 31
2020-06-01 2020-06-30 30
2020-06-01 2020-06-30 30
2020-07-01 2020-07-31 31
2020-07-01 2020-07-31 31
2020-08-01 2020-08-31 31
2020-08-01 2020-08-31 31
2020-09-01 2020-09-30 30
2020-09-01 2020-09-30 30
2020-10-01 2020-10-31 31
2020-10-01 2020-10-31 31
2020-11-01 2020-11-30 30
2020-11-01 2020-11-30 30
2020-12-01 2020-12-31 31
2020-12-01 2020-12-31 31


In [194]:
day1 = datetime.date(2021,10,18)
day2 = datetime.date(2021,10,5)
print(type((day2-day1).days))

<class 'int'>


This notebook is part of a python assessment for a knowledge check for a course at www.codeimmersives.com . If you accessed this notebook outside the course, you can get more information about this course online by 
clicking [here](https://www.codeimmersives.com/programs/python-aws/).

<hr>

Copyright &copy; 2021  Code Immersives